src/dfcx_scrapi/tools/test_cases_util.py

```diff
@@ -0,0 +1,559 @@
+                    return None
+            new_conversation_turns.append(conv_turn)
+
+        return new_conversation_turns
+
+    def migrate_test_cases(
+        self,
+        source_agent: str,
+        target_agent: str,
+        rate_limit: int = 5) -> List[types.TestCase]:
+        """The purpose of this method is to create a new set of the test cases
+            from (old) a source agent to (new) a target agent.
+            When the new agent (target) is created by the copy util,
+            UUIDs of flows, pages, intents, and etc are newly generated.
+            Therefore, the discrepencies of the uuids creates conflict
+            when importing the test cases from a source to a target agent.
+            To prevent this, this method helps to migrate the test cases by
+            replacing the UUIDs that are relevant to a target agent. The test
+            cases are only migratable if the display names of flows, pages,
+            intents, and etc have the matching set in both source and target
+            agents.
+          Args:
+            source_agent: The agent that contain the test cases for.
+              Format:
+                `projects/<ProjectID>/locations/<LocationID>/agents/<AgentID>`
+            target_agent: The new agent where you want to migrate.
+              Format:
+                `projects/<ProjectID>/locations/<Location ID>/agents/<AgentID>`
+          Returns:
+            a list of the types.TestCase objects that are created in the
+            target agent.
+        """
+
+        source_dfcx_test_cases = test_cases.TestCases(
+            creds=self.creds, agent_id=source_agent)
+        target_dfcx_test_cases = test_cases.TestCases(
+            creds=self.creds, agent_id=target_agent)
+        source_agent_tcs = source_dfcx_test_cases.list_test_cases(
+            source_agent, include_conversation_turns=True)
+        if not source_agent_tcs:
+            raise f"source agent:{source_agent} does not have any test cases"
+        new_test_cases = []
+        s_commons = self._get_commons_config(agent_id=source_agent)
+        t_commons = self._get_commons_config(agent_id=target_agent)
+        for tc in source_agent_tcs:
+            new_test_config = self._get_new_test_config(
+                s_commons, t_commons, tc)
+            new_conv_turns = self._get_new_conversation_turns(
+                s_commons, t_commons, tc.test_case_conversation_turns)
+            new_last_test_conv_turns = self._get_new_conversation_turns(
+                s_commons, t_commons, tc.last_test_result.conversation_turns)
+            if None in [
+                new_test_config,
+                new_conv_turns,
+                new_last_test_conv_turns]:
+                if not new_test_config:
+                    logging.warning(
+                        f"test_case: {tc.display_name} "\
+                        f"Reason: test_config is None.")
```

Comment:
You don't need `\` at the end. Strings will concatenate anyway and also you don't need the second string to use f string formats.
```
                    logging.warning(
                        f"test_case: {tc.display_name} "
                        "Reason: test_config is None.")
```
with that being said, for logging messages you need to use [this format](https://google.github.io/styleguide/pyguide.html#3101-logging)

Reply:
ok, fixed